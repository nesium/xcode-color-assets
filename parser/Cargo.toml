[package]
name = "parser"
version = "0.1.0"
authors = ["mb <mb@nesium.com>"]
edition = "2018"

[dependencies]
nom = "5.0"

[dev-dependencies]
insta = "0.10"