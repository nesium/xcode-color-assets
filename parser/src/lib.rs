pub mod ast;
mod parser;
mod var_context;
mod error;

pub use self::parser::{parse_document, parse_document_from_file};
pub use self::var_context::{VarContext, ResolvedVariable, ResolvedColorSet};
pub use self::error::Error;
