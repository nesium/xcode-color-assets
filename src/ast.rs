#[derive(Debug, PartialEq)]
pub struct Document {
  pub items: Vec<DocumentItem>,
}

#[derive(Debug, PartialEq)]
pub enum DocumentItem {
  Variable(Declaration<Value>),
  RuleSet(RuleSet),
}

#[derive(Debug, Clone, PartialEq)]
pub struct Color {
  pub r: u8,
  pub g: u8,
  pub b: u8,
  pub a: f32,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Value {
  Variable(String),
  Color(Color),
  ColorSet(ColorSet),
}

#[derive(Debug, Clone, PartialEq)]
pub enum ColorSetValue {
  Variable(String),
  Color(Color),
}

#[derive(Debug, Clone, PartialEq)]
pub struct ColorSet {
  pub light: ColorSetValue,
  pub dark: ColorSetValue,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Declaration<T> {
  pub identifier: String,
  pub value: T,
}

#[derive(Debug, Clone, PartialEq)]
pub struct RuleSet {
  pub identifier: String,
  pub items: Vec<RuleSetItem>,
}

#[derive(Debug, Clone, PartialEq)]
pub enum RuleSetItem {
  RuleSet(RuleSet),
  Declaration(Declaration<Value>),
}

impl Into<Option<Color>> for Value {
  fn into(self) -> Option<Color> {
    match self {
      Value::Color(c) => Some(c),
      _ => None,
    }
  }
}

impl Into<Value> for ColorSetValue {
  fn into(self) -> Value {
    match self {
      ColorSetValue::Color(c) => Value::Color(c),
      ColorSetValue::Variable(v) => Value::Variable(v),
    }
  }
}
