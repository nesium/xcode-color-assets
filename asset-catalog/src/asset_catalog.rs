use super::Error;
use parser::ast::{
  Color, ColorSet, ColorSetValue, Declaration, Document, DocumentItem, RuleSet, RuleSetItem, Value,
};
use serde_json::json;
use std::collections::HashMap;
use std::fs;
use std::path::Path;

fn path_to_str(path: &Path) -> String {
  path
    .file_name()
    .and_then(|s| s.to_str())
    .map(String::from)
    .unwrap()
}

pub fn write_asset_catalog(
  doc: &Document,
  path: &Path,
  delete_directory_if_exists: bool,
) -> Result<(), Error> {
  let var_lookup: HashMap<String, &Value> = doc
    .items
    .iter()
    .filter_map(|item| match item {
      DocumentItem::Variable(v) => Some(v),
      _ => None,
    })
    .map(|variable| (variable.identifier.to_string(), &variable.value))
    .collect();

  if path.exists() {
    if delete_directory_if_exists {
      fs::remove_dir_all(&path).map_err(|_| Error::CouldNotRemoveDirectory(path_to_str(&path)))?;
    } else {
      return Err(Error::CatalogExists(path_to_str(path)));
    }
  }
  fs::create_dir_all(&path).map_err(|_| Error::CouldNotCreateDirectory(path_to_str(&path)))?;

  for item in doc.items.iter() {
    match item {
      DocumentItem::RuleSet(r) => {
        write_ruleset(r, &path, &r.identifier, &var_lookup)?;
      }
      DocumentItem::Declaration(d) => {
        write_declaration(d, &path, &d.identifier, &var_lookup)?;
      }
      DocumentItem::Variable(_) => {}
    }
  }

  Ok(())
}

fn write_ruleset(
  ruleset: &RuleSet,
  path: &Path,
  identifier: &str,
  var_lookup: &HashMap<String, &Value>,
) -> Result<(), Error> {
  let ruleset_path = path.join(&ruleset.identifier);
  fs::create_dir(&ruleset_path)
    .map_err(|_| Error::CouldNotCreateDirectory(path_to_str(&ruleset_path)))?;

  let info = json!({
    "info": {
      "version": 1,
      "author": "xcode"
    }
  });

  let json_path = ruleset_path.join("Contents.json");

  fs::write(
    &json_path,
    serde_json::to_string_pretty(&info).unwrap().as_bytes(),
  )
  .map_err(|_| Error::CouldNotCreateFile(path_to_str(&json_path)))?;

  for item in ruleset.items.iter() {
    match item {
      RuleSetItem::RuleSet(r) => {
        let child_identifier = format!("{}{}", identifier, r.identifier);
        write_ruleset(r, &ruleset_path, &child_identifier, &var_lookup)?;
      }
      RuleSetItem::Declaration(d) => {
        let child_identifier = format!("{}{}", identifier, d.identifier);
        write_declaration(d, &ruleset_path, &child_identifier, &var_lookup)?;
      }
    }
  }

  Ok(())
}

fn write_declaration(
  declaration: &Declaration<Value>,
  path: &Path,
  identifier: &str,
  var_lookup: &HashMap<String, &Value>,
) -> Result<(), Error> {
  let colorset_path = path.join(identifier.to_string()).with_extension("colorset");

  fs::create_dir(&colorset_path)
    .map_err(|_| Error::CouldNotCreateDirectory(path_to_str(&colorset_path)))?;

  enum ResolvedVariable {
    Color(Color),
    ColorSet(ColorSet),
  }

  fn resolve_variable<F>(identifier: &str, var_lookup: &HashMap<String, &Value>, mut resolver: F)
  where
    F: FnMut(&ResolvedVariable) -> (),
  {
    let value = var_lookup
      .get(identifier)
      .expect(&format!("Referenced undefined variable {}", identifier));
    match value {
      Value::Variable(identifier) => resolve_variable(identifier, var_lookup, resolver),
      Value::Color(color) => resolver(&ResolvedVariable::Color(color.clone())),
      Value::ColorSet(colorset) => resolver(&ResolvedVariable::ColorSet(colorset.clone())),
    }
  };

  let mut info: serde_json::value::Value = json!({
    "info": {
      "version": 1,
      "author": "xcode"
    },
    "colors": []
  });

  let components = |color: &Color| {
    json!({
      "red": format!("0x{:02X}", color.r),
      "green": format!("0x{:02X}", color.g),
      "blue": format!("0x{:02X}", color.b),
      "alpha": format!("{:.3}", color.a)
    })
  };

  let append_light_color = |value: &mut serde_json::value::Value, color: &Color| {
    let arr = value["colors"].as_array_mut().unwrap();
    arr.push(json!({
      "idiom": "universal",
      "color": {
        "color-space" : "srgb",
        "components": components(color)
      }
    }));
  };

  let append_dark_color = |value: &mut serde_json::value::Value, color: &Color| {
    let arr = value["colors"].as_array_mut().unwrap();
    arr.push(json!({
      "idiom" : "universal",
      "appearances" : [{
        "appearance" : "luminosity",
        "value" : "dark"
      }],
      "color" : {
        "color-space" : "srgb",
        "components" : components(color)
      }
    }));
  };

  let append_colorset = |value: &mut serde_json::value::Value, colorset: &ColorSet| {
    match colorset.light {
      ColorSetValue::Variable(ref identifier) => {
        resolve_variable(identifier, var_lookup, |result| match result {
          ResolvedVariable::Color(color) => append_light_color(value, color),
          ResolvedVariable::ColorSet(_) => panic!(format!(
            "Attempt to assign a colorset to the light value of another colorset via variable {}",
            identifier
          )),
        });
      }
      ColorSetValue::Color(ref color) => append_light_color(value, color),
    }

    match colorset.dark {
      ColorSetValue::Variable(ref identifier) => {
        resolve_variable(identifier, var_lookup, |result| match result {
          ResolvedVariable::Color(color) => append_dark_color(value, color),
          ResolvedVariable::ColorSet(_) => panic!(format!(
            "Attempt to assign a colorset to the dark value of another colorset via variable {}",
            identifier
          )),
        });
      }
      ColorSetValue::Color(ref color) => append_dark_color(value, color),
    }
  };

  match declaration.value {
    Value::Variable(ref identifier) => {
      resolve_variable(identifier, var_lookup, |result| match result {
        ResolvedVariable::Color(color) => append_light_color(&mut info, color),
        ResolvedVariable::ColorSet(colorset) => append_colorset(&mut info, colorset),
      });
    }
    Value::Color(ref color) => append_light_color(&mut info, color),
    Value::ColorSet(ref colorset) => append_colorset(&mut info, colorset),
  }

  let json_path = colorset_path.join("Contents.json");

  fs::write(
    &json_path,
    serde_json::to_string_pretty(&info).unwrap().as_bytes(),
  )
  .map_err(|_| Error::CouldNotCreateFile(path_to_str(&json_path)))
}
