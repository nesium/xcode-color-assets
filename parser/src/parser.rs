use super::ast::{
  Color, ColorSet, ColorSetValue, Declaration, Document, DocumentItem, RuleSet, RuleSetItem, Value,
};
use nom::branch::alt;
use nom::bytes::complete::{tag, take_while_m_n};
use nom::character::complete::{alphanumeric1, char, digit1, multispace0, newline, space0, space1};
use nom::combinator::{all_consuming, cut, map, map_res, opt};
use nom::error::{context, convert_error, ParseError, VerboseError};
use nom::multi::{many0, separated_list};
use nom::number::complete::float;
use nom::sequence::{delimited, preceded, separated_pair, terminated, tuple};
use nom::Err;
use nom::IResult;

use std::fmt;
use std::fs::File;
use std::io::prelude::Read;
use std::str;

#[derive(Debug)]
pub struct Error {
  error: String,
}

impl std::error::Error for Error {
  fn description(&self) -> &str {
    &self.error
  }
}

impl fmt::Display for Error {
  fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
    write!(f, "Error: {}", self.error)
  }
}

fn parse_hex_value(input: &str) -> Result<u32, std::num::ParseIntError> {
  u32::from_str_radix(input, 16)
}

fn colorset_from_declarations(
  decl1: Declaration<ColorSetValue>,
  decl2: Declaration<ColorSetValue>,
) -> Result<ColorSet, Error> {
  match (decl1.identifier.as_str(), decl2.identifier.as_str()) {
    ("light", "dark") => Ok(ColorSet {
      light: decl1.value,
      dark: decl2.value,
    }),
    ("dark", "light") => Ok(ColorSet {
      light: decl2.value,
      dark: decl1.value,
    }),
    _ => Err(Error {
      error: format!(
        "Expected light & dark properties. Found {}, {}.",
        decl1.identifier, decl2.identifier
      ),
    }),
  }
}

fn line_delimiter<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, char, E> {
  cut(terminated(preceded(space0, newline), multispace0))(input)
}

pub fn parse_document(input: String) -> Result<Document, Error> {
  let mut modified_input = input;
  modified_input.push_str("\n");

  let result: IResult<&str, Document, VerboseError<&str>> = map(
    all_consuming(delimited(
      multispace0,
      separated_list(
        line_delimiter,
        alt((
          map(variable, DocumentItem::Variable),
          map(ruleset, DocumentItem::RuleSet),
          map(declaration(value), DocumentItem::Declaration),
        )),
      ),
      multispace0,
    )),
    |res| Document { items: res },
  )(&modified_input);

  match result {
    Ok((_, doc)) => Ok(doc),
    Err(Err::Error(e)) | Err(Err::Failure(e)) => Err(Error {
      error: convert_error(&modified_input, e).replace("'\n'", "'\\n'"),
    }),
    _ => {
      eprintln!("An unknown error occurred.");
      std::process::exit(0x0100)
    }
  }
}

pub fn parse_document_from_file(filepath: &str) -> Result<Document, Error> {
  let mut f = File::open(filepath).unwrap();
  let mut buffer = vec![];
  f.read_to_end(&mut buffer).unwrap();

  let contents = match str::from_utf8(&buffer) {
    Ok(v) => v,
    Err(e) => panic!(
      "Could not read contents of file {}. Reason: {}",
      filepath, e
    ),
  };

  parse_document(contents.to_string())
}

pub fn ruleset<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, RuleSet, E> {
  let body = |input| {
    delimited(
      terminated(char('{'), multispace0),
      cut(many0(terminated(
        alt((
          map(ruleset, RuleSetItem::RuleSet),
          map(declaration(value), RuleSetItem::Declaration),
        )),
        line_delimiter,
      ))),
      preceded(multispace0, char('}')),
    )(input)
  };

  context(
    "RuleSet",
    map(separated_pair(identifier, space0, body), |res| RuleSet {
      identifier: res.0.to_string(),
      items: res.1,
    }),
  )(input)
}

pub fn variable<'a, E: ParseError<&'a str>>(
  input: &'a str,
) -> IResult<&'a str, Declaration<Value>, E> {
  context(
    "Variable",
    map(
      separated_pair(
        variable_identifier,
        cut(terminated(preceded(space0, char(':')), space0)),
        value,
      ),
      |res| {
        let (identifier, value) = res;
        Declaration { identifier, value }
      },
    ),
  )(input)
}

fn value<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, Value, E> {
  preceded(
    space0,
    alt((
      map(colorset_value, |res| res.into()),
      map(colorset, Value::ColorSet),
    )),
  )(input)
}

fn colorset_value<'a, E: ParseError<&'a str>>(
  input: &'a str,
) -> IResult<&'a str, ColorSetValue, E> {
  preceded(
    space0,
    alt((
      map(hex_color, ColorSetValue::Color),
      map(rgba_color, ColorSetValue::Color),
      map(variable_identifier, ColorSetValue::Variable),
    )),
  )(input)
}

fn rgba_color<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, Color, E> {
  let u8_value = move |input: &'a str| {
    preceded(
      space0,
      cut(context(
        "Value between 0 - 255",
        map_res(digit1, |s: &str| s.parse::<u8>()),
      )),
    )(input)
  };
  let delimiter = move |input: &'a str| preceded(space0, cut(char(',')))(input);

  context(
    "RGBA Value",
    map(
      preceded(
        tag("rgba"),
        delimited(
          preceded(space0, char('(')),
          tuple((
            terminated(u8_value, delimiter),
            terminated(u8_value, delimiter),
            terminated(u8_value, delimiter),
            preceded(space0, float),
          )),
          preceded(space0, char(')')),
        ),
      ),
      |res| {
        let (r, g, b, a) = res;
        Color { r, g, b, a }
      },
    ),
  )(input)
}

fn hex_color<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, Color, E> {
  map(
    tuple((
      char('#'),
      cut(hex_value),
      opt(preceded(space1, alpha_value)),
    )),
    |res| {
      let (_, rgb, alpha) = res;
      let r = ((rgb >> 16) & 0xff) as u8;
      let g = ((rgb >> 8) & 0xff) as u8;
      let b = (rgb & 0xff) as u8;
      let a = alpha.unwrap_or(1.0);
      Color { r, g, b, a }
    },
  )(input)
}

pub fn colorset<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, ColorSet, E> {
  context(
    "ColorSet",
    map_res(
      delimited(
        cut(char('(')),
        separated_pair(
          cut(declaration(colorset_value)),
          cut(preceded(space0, char(','))),
          cut(declaration(colorset_value)),
        ),
        cut(char(')')),
      ),
      |res| colorset_from_declarations(res.0, res.1),
    ),
  )(input)
}

pub fn identifier<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, String, E> {
  context(
    "Identifier",
    map(preceded(space0, alphanumeric1), |ident: &str| {
      ident.to_string()
    }),
  )(input)
}

pub fn variable_identifier<'a, E: ParseError<&'a str>>(
  input: &'a str,
) -> IResult<&'a str, String, E> {
  context(
    "Variable Identifier",
    map(
      preceded(space0, preceded(char('$'), alphanumeric1)),
      |ident: &str| ident.to_string(),
    ),
  )(input)
}

pub fn declaration<'a, F, V, E: ParseError<&'a str>>(
  value: F,
) -> impl Fn(&'a str) -> IResult<&'a str, Declaration<V>, E>
where
  F: Fn(&'a str) -> IResult<&'a str, V, E>,
{
  move |input: &'a str| {
    context(
      "Declaration",
      map(
        separated_pair(identifier, cut(preceded(space0, char(':'))), &value),
        |res| {
          let (identifier, value) = res;
          Declaration { identifier, value }
        },
      ),
    )(input)
  }
}

fn hex_value<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, u32, E> {
  context(
    "Hex Value",
    map_res(take_while_m_n(6, 6, is_hex_digit), parse_hex_value),
  )(input)
}

fn alpha_value<'a, E: ParseError<&'a str>>(input: &'a str) -> IResult<&'a str, f32, E> {
  context(
    "Alpha Value",
    map_res(terminated(digit1, tag("%")), |input| {
      u32::from_str_radix(input, 10).map(|val| (val as f32) / 100.0)
    }),
  )(input)
}

fn is_hex_digit(c: char) -> bool {
  c.is_digit(16)
}
