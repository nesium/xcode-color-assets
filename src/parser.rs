use super::ast::{
  Color, ColorSet, ColorSetValue, Declaration, Document, DocumentItem, RuleSet, RuleSetItem, Value,
};
use nom::branch::alt;
use nom::bytes::complete::{tag, take_while_m_n};
use nom::character::complete::{alphanumeric1, char, digit1, multispace0, newline, space0, space1};
use nom::combinator::{map, map_res, opt};
use nom::error::{context, ParseError, VerboseError};
use nom::multi::{many0, separated_list};
use nom::sequence::{preceded, separated_pair, terminated, tuple};
use nom::Err;
use nom::IResult;

use std::fs::File;
use std::io::prelude::Read;
use std::str;

#[allow(dead_code)]
fn parse_hex_value(input: &str) -> Result<u32, std::num::ParseIntError> {
  u32::from_str_radix(input, 16)
}

#[allow(dead_code)]
fn colorset_from_declarations(
  decl1: Declaration<ColorSetValue>,
  decl2: Declaration<ColorSetValue>,
) -> Result<ColorSet, String> {
  match (decl1.identifier.as_str(), decl2.identifier.as_str()) {
    ("light", "dark") => Ok(ColorSet {
      light: decl1.value,
      dark: decl2.value,
    }),
    ("dark", "light") => Ok(ColorSet {
      light: decl2.value,
      dark: decl1.value,
    }),
    _ => Err(format!(
      "Expected light & dark properties. Found {}, {}.",
      decl1.identifier, decl2.identifier
    )),
  }
}

#[allow(dead_code)]
fn line_delimiter(input: &str) -> IResult<&str, char> {
  terminated(preceded(space0, newline), multispace0)(input)
}

#[allow(dead_code)]
pub fn parse_document(input: &str) -> IResult<&str, Document> {
  map(
    preceded(
      multispace0,
      separated_list(
        line_delimiter,
        alt((
          map(variable, DocumentItem::Variable),
          map(ruleset, DocumentItem::RuleSet),
        )),
      ),
    ),
    |res| Document { items: res },
  )(input)
}

pub fn parse_document_from_file(filepath: &str) -> Result<Document, VerboseError<String>> {
  let mut f = File::open(filepath).unwrap();
  let mut buffer = vec![];
  f.read_to_end(&mut buffer).unwrap();

  let contents = match str::from_utf8(&buffer) {
    Ok(v) => v,
    Err(e) => panic!(
      "Could not read contents of file {}. Reason: {}",
      filepath, e
    ),
  };

  match parse_document(&contents) {
    Ok(res) => Ok(res.1),
    Err(Err::Error(e)) | Err(Err::Failure(e)) => {
      Err(VerboseError::from_error_kind(contents.to_string(), e.1))
    }
    _ => {
      eprintln!("Something went wrong");
      std::process::exit(0x0100)
    }
  }
}

#[allow(dead_code)]
pub fn ruleset(input: &str) -> IResult<&str, RuleSet> {
  let body = |input| {
    preceded(
      terminated(char('{'), multispace0),
      terminated(
        many0(alt((
          map(terminated(ruleset, line_delimiter), RuleSetItem::RuleSet),
          map(
            terminated(declaration(value), line_delimiter),
            RuleSetItem::Declaration,
          ),
        ))),
        preceded(multispace0, char('}')),
      ),
    )(input)
  };

  context(
    "ruleset",
    map(tuple((identifier, space0, body)), |res| RuleSet {
      identifier: res.0.to_string(),
      items: res.2,
    }),
  )(input)
}

pub fn variable(input: &str) -> IResult<&str, Declaration<Value>> {
  map(
    separated_pair(
      identifier,
      terminated(preceded(space0, char('=')), space0),
      value,
    ),
    |res| {
      let (identifier, value) = res;
      Declaration { identifier, value }
    },
  )(input)
}

fn value(input: &str) -> IResult<&str, Value> {
  alt((
    map(colorset_value, |res| res.into()),
    map(colorset, Value::ColorSet),
  ))(input)
}

fn colorset_value(input: &str) -> IResult<&str, ColorSetValue> {
  alt((
    map(hex_color, ColorSetValue::Color),
    map(preceded(char('$'), identifier), ColorSetValue::Variable),
  ))(input)
}

fn hex_color(input: &str) -> IResult<&str, Color> {
  map(
    tuple((tag("#"), hex_value, opt(preceded(space1, alpha_value)))),
    |res| {
      let (_, rgb, alpha) = res;
      let r = ((rgb >> 16) & 0xff) as u8;
      let g = ((rgb >> 8) & 0xff) as u8;
      let b = (rgb & 0xff) as u8;
      let a = alpha.unwrap_or(1.0);
      Color { r, g, b, a }
    },
  )(input)
}

pub fn colorset(input: &str) -> IResult<&str, ColorSet> {
  map_res(
    tuple((
      tag("("),
      declaration(colorset_value),
      space0,
      tag(","),
      multispace0,
      declaration(colorset_value),
      tag(")"),
    )),
    |res| colorset_from_declarations(res.1, res.5),
  )(input)
}

pub fn identifier(input: &str) -> IResult<&str, String> {
  map(preceded(space0, alphanumeric1), |ident: &str| {
    ident.to_string()
  })(input)
}

pub fn declaration<F, V>(value: F) -> impl Fn(&str) -> IResult<&str, Declaration<V>>
where
  F: Fn(&str) -> IResult<&str, V>,
{
  move |input: &str| {
    map(
      separated_pair(
        identifier,
        terminated(preceded(space0, char(':')), space0),
        &value,
      ),
      |res| {
        let (identifier, value) = res;
        Declaration { identifier, value }
      },
    )(input)
  }
}

fn hex_value(input: &str) -> IResult<&str, u32> {
  map_res(take_while_m_n(6, 6, is_hex_digit), parse_hex_value)(input)
}

fn alpha_value(input: &str) -> IResult<&str, f32> {
  map_res(terminated(digit1, tag("%")), |input| {
    u32::from_str_radix(input, 10).map(|val| (val as f32) / 100.0)
  })(input)
}

fn is_hex_digit(c: char) -> bool {
  c.is_digit(16)
}
