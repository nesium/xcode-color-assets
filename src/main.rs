mod asset_catalog;
mod ast;
mod parser;

use crate::asset_catalog::write_asset_catalog;
use crate::parser::parse_document_from_file;
use clap::{App, Arg};

fn main() {
  let matches = App::new("color-assets")
    .version("1.0")
    .about("Create Xcode Asset Catalog with colors for light & dark mode.")
    .arg(
      Arg::with_name("output")
        .short("o")
        .help("Sets the output filename")
        .value_name("OUTPUT_FILE")
        .required(true),
    )
    .arg(
      Arg::with_name("input")
        .help("Sets the input file")
        .value_name("INPUT_FILE")
        .required(true)
        .index(1),
    )
    .get_matches();

  let input_file = matches.value_of("input").unwrap();
  let output_path = matches.value_of("output").unwrap();

  let doc = parse_document_from_file(&input_file).expect("Could not parse input file.");
  write_asset_catalog(&doc, &output_path).expect("Could not write asset catalog.");
}

#[cfg(test)]
mod tests {
  use crate::ast::{Color, ColorSet, ColorSetValue, Value};
  use crate::parser::{colorset, parse_document, ruleset, variable};
  use insta::assert_debug_snapshot_matches;

  #[test]
  fn test_document() {
    let doc = r#"
    mediumBright = #aabbcc 33%
    red = #ff0000
    white = #ffffff

    root {
      ApplicationBackground: (light: $white, dark: #141517)

      NumericInput {
        ActionKey {
          Background: (light: $red, dark: #ff00ff)
          Highlight: (light: #cccccc, dark: #000000)
        }
      }
    }
    "#;

    let (_, doc) = parse_document(&doc).unwrap();
    assert_debug_snapshot_matches!("test_document", doc);
  }

  #[test]
  fn test_ruleset() {
    let doc = r#"root {
      ApplicationBackground: (light: $white, dark: #141517)

      Key {
        Background: (light: $red, dark: #ff00ff)
      }

      AnotherKey {
        Highlight: (light: #cccccc, dark: #000000)
      }
    }"#;

    let (_, doc) = ruleset(&doc).unwrap();
    assert_debug_snapshot_matches!("test_ruleset", doc);
  }

  #[test]
  fn test_color_variable() {
    let (_, v1) = variable(&"myColor = #ff00ff 44%").unwrap();
    assert_eq!(v1.identifier, "myColor");
    assert_eq!(
      v1.value,
      Value::Color(Color {
        r: 255,
        g: 0,
        b: 255,
        a: 0.44
      })
    );

    let (_, v2) = variable(&"a = #4224be").unwrap();
    assert_eq!(v2.identifier, "a");
    assert_eq!(
      v2.value,
      Value::Color(Color {
        r: 66,
        g: 36,
        b: 190,
        a: 1.0
      })
    );

    let (_, v3) = variable(&"a1 = #4B0FC6").unwrap();
    assert_eq!(v3.identifier, "a1");
    assert_eq!(
      v3.value,
      Value::Color(Color {
        r: 75,
        g: 15,
        b: 198,
        a: 1.0
      })
    );
  }

  #[test]
  fn test_colorset() {
    let val1 = ColorSetValue::Color(Color {
      r: 255,
      g: 0,
      b: 255,
      a: 0.3,
    });
    let val2 = ColorSetValue::Color(Color {
      r: 0,
      g: 255,
      b: 0,
      a: 1.0,
    });

    let expected_result = ColorSet {
      light: val1.clone(),
      dark: val2,
    };

    let (_, v1) = colorset(&"(light: #ff00ff 30%, dark: #00ff00)").unwrap();
    assert_eq!(v1, expected_result);
    let (_, v2) = colorset(&"(dark: #00ff00, light: #ff00ff 30%)").unwrap();
    assert_eq!(v2, expected_result);
    let (_, v3) = colorset(&"(dark: $applicationBackgroundLight, light: #ff00ff 30%)").unwrap();
    assert_eq!(
      v3,
      ColorSet {
        light: val1,
        dark: ColorSetValue::Variable("applicationBackgroundLight".to_string()),
      }
    );
  }
}
