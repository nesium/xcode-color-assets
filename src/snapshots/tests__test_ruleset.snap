---
created: "2019-08-15T12:40:09.784581Z"
creator: insta@0.10.1
source: src/main.rs
expression: doc
---
RuleSet {
    identifier: "root",
    items: [
        Declaration(
            Declaration {
                identifier: "ApplicationBackground",
                value: ColorSet(
                    ColorSet {
                        light: Variable(
                            "white",
                        ),
                        dark: Color(
                            Color {
                                r: 20,
                                g: 21,
                                b: 23,
                                a: 1.0,
                            },
                        ),
                    },
                ),
            },
        ),
        RuleSet(
            RuleSet {
                identifier: "Key",
                items: [
                    Declaration(
                        Declaration {
                            identifier: "Background",
                            value: ColorSet(
                                ColorSet {
                                    light: Variable(
                                        "red",
                                    ),
                                    dark: Color(
                                        Color {
                                            r: 255,
                                            g: 0,
                                            b: 255,
                                            a: 1.0,
                                        },
                                    ),
                                },
                            ),
                        },
                    ),
                ],
            },
        ),
        RuleSet(
            RuleSet {
                identifier: "AnotherKey",
                items: [
                    Declaration(
                        Declaration {
                            identifier: "Highlight",
                            value: ColorSet(
                                ColorSet {
                                    light: Color(
                                        Color {
                                            r: 204,
                                            g: 204,
                                            b: 204,
                                            a: 1.0,
                                        },
                                    ),
                                    dark: Color(
                                        Color {
                                            r: 0,
                                            g: 0,
                                            b: 0,
                                            a: 1.0,
                                        },
                                    ),
                                },
                            ),
                        },
                    ),
                ],
            },
        ),
    ],
}
